### PR TITLE
spotbugs 3.1.0 (new formula)

### PR DESCRIPTION
**Pull request by stephengroat**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
since it's primarily a GUI app, i thought the test was ok since it exercises how the app and the JAR are running on the command line

### CODE REVIEW

File: Formula/spotbugs.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Spotbugs < Formula
+  desc "SpotBugs is FindBugs' successor. A tool for static analysis to look for bugs in Java code."
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: Let's keep it to one sentence, and no period at the end

File: Formula/spotbugs.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  def install
+    system "gradle build"
+    system "gradle installDist"
+    libexec.install Dir["spotbugs/build/install/spotbugs/*"]
+    bin.install_symlink "#{libexec}/bin/spotbugs"
+  end
+
+  test do
+    assert_match "3.1.0", shell_output("#{bin}/spotbugs -version", 1)
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: The test needs to exercise and verify the functionality of the software.

File: Formula/spotbugs.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  head "https://github.com/spotbugs/spotbugs.git"
+
+  depends_on "gradle" => :build
+  depends_on :java => "1.8+"
+
+  def install
+    system "gradle build"
+    system "gradle installDist"
+    libexec.install Dir["spotbugs/build/install/spotbugs/*"]
+    bin.install_symlink "#{libexec}/bin/spotbugs"
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: any reason none of the others in libexec/bin?

**Review by ilovezfs — CHANGES_REQUESTED** [review]

**Comment by ilovezfs** [comment]
Also this has a whole pile of Java 9 warnings that need to be reported upstream as they will be errors, not warnings, in Java 10.

File: Formula/spotbugs.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  homepage "https://spotbugs.github.io/"
+  url "https://github.com/spotbugs/spotbugs/archive/3.1.0.tar.gz"
+  sha256 "ac6169a551212756a05dd9c066ef9212653d8a99b754ef4dade6a7008f8ad781"
+  head "https://github.com/spotbugs/spotbugs.git"
+
+  depends_on "gradle" => :build
+  depends_on :java => "1.8+"
+
+  def install
+    system "gradle build"
```
Comment (bfontaine) [inline Formula/spotbugs.rb]: `system "gradle", "build"`

File: Formula/spotbugs.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    system "gradle installDist"
```
Comment (bfontaine) [inline Formula/spotbugs.rb]: `system "gradle", "installDist"`

**Review by bfontaine — CHANGES_REQUESTED** [review]

File: Formula/spotbugs.rb (hunk 2 above)
Comment (stephengroat) [reply Formula/spotbugs.rb]: done

File: Formula/spotbugs.rb (hunk 3 above)
Comment (stephengroat) [reply Formula/spotbugs.rb]: `spotbugs` is a wrapper that sets some other variables, etc. they may work depending on system configuration, but `spotbugs` is the best internally curated binary

File: Formula/spotbugs.rb (hunk 1 above)
Comment (stephengroat) [reply Formula/spotbugs.rb]: done

**Comment by stephengroat** [comment]
https://github.com/Homebrew/homebrew-core/pull/19886#issuecomment-339663748 the project is aware and currently working on fixing the java 1.9 issues

File: Formula/spotbugs.rb (hunk 4 above)
Comment (stephengroat) [reply Formula/spotbugs.rb]: done

File: Formula/spotbugs.rb (hunk 5 above)
Comment (stephengroat) [reply Formula/spotbugs.rb]: done

File: Formula/spotbugs.rb [hunk 6]
```diff
@@ -16,6 +16,18 @@ def install
...
+      private double[] myList;
+      public static void main(String[] args) {
+        System.out.println("Hello World");
+      }
+      public double[] getList() {
+        return myList;
+      }
+    }
+    EOS
+    system "javac HelloWorld.java"
```
Comment (bfontaine) [inline Formula/spotbugs.rb]: `system "javac", "HelloWorld.java"`

File: Formula/spotbugs.rb [hunk 7]
```diff
@@ -16,6 +16,18 @@ def install
...
+    system "jar cvfe HelloWorld.jar HelloWorld *.class"
+    assert_match /M V EI.*\nM C UwF.*\n/, shell_output("#{bin}/spotbugs -textui HelloWorld.jar")  end
```
Comment (bfontaine) [inline Formula/spotbugs.rb]: Please add a newline before `end`.

**Review by bfontaine — CHANGES_REQUESTED** [review]

**Comment by stephengroat** [comment]
since this is `findbugs` successor, is there any way to remove that formula as well?

**Comment by dunn** [comment]
Yeah, feel free to add a commit removing that as well.

**Comment by ilovezfs** [comment]
>since this is findbugs successor

do you have a reference for this? It's possible we should actually treat this as a rename.

**Comment by ilovezfs** [comment]
yup let's treat it as a rename. so you'll also need an alias and an entry in formula_renames.json

**Review by ilovezfs — DISMISSED** [review]

**Review by ilovezfs — CHANGES_REQUESTED** [review]
It looks like upstream provides prebuilt jars so we should use those and make it bottle unneeded.

**Comment by ilovezfs** [comment]
http://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz

**Review by ilovezfs — APPROVED** [review]

File: Formula/spotbugs.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+          System.out.println("Hello World");
+        }
+        public double[] getList() {
+          return myList;
+        }
+      }
+    EOS
+    system "javac", "HelloWorld.java"
+    system "jar", "cvfe", "HelloWorld.jar", "HelloWorld", "HelloWorld.class"
+    assert_match /M V EI.*\nM C UwF.*\n/, shell_output("#{bin}/spotbugs -textui HelloWorld.jar")
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: This needs to be split over two lines since it exceeds the eightieth column.

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/spotbugs.rb (hunk 8 above)
Comment (ilovezfs) [reply Formula/spotbugs.rb]:
```
    output = shell_output("#{bin}/spotbugs -textui HelloWorld.jar")
    assert_match /M V EI.*M C UwF/m, output
```

Comment (stephengroat) [reply Formula/spotbugs.rb]: addressed

Comment (stephengroat) [reply Formula/spotbugs.rb]: i always turn on the `Allow edits from maintainers.` button, please feel free to edit my submissions as well

File: Formula/spotbugs.rb [hunk 9]
```diff
@@ -35,6 +35,7 @@ def install
     EOS
     system "javac", "HelloWorld.java"
     system "jar", "cvfe", "HelloWorld.jar", "HelloWorld", "HelloWorld.class"
-    assert_match /M V EI.*\nM C UwF.*\n/, shell_output("#{bin}/spotbugs -textui HelloWorld.jar")
+    output = shell_output("#{bin}/spotbugs -textui HelloWorld.jar")
+    assert_match /M V EI.*\nM C UwF.*\n/, output
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: you can avoid the \n's by using /m at the end.

```
assert_match /M V EI.*M C UwF/m, output
```
should work

Comment (stephengroat) [reply Formula/spotbugs.rb]: Since each bug shows up on a separate line, I specifically wanted to catch the fact the 2 "bugs" were found (if that's ok)

Comment (ilovezfs) [reply Formula/spotbugs.rb]: sure that's fine.

**Review by ilovezfs — APPROVED** [review]

File: Formula/spotbugs.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+class Spotbugs < Formula
+  desc "Tool for Java static analysis (FindBugs' successor)"
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: Technically this should be `FindBugs's`

Comment (stephengroat) [reply Formula/spotbugs.rb]: that's quality grammer catching. fixed

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
```
Error: 2 problems in 1 formula
spotbugs:
  * C: 4: col 3: Please use https:// for http://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz
  * Stable: The URL http://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz should use HTTPS rather than HTTP
```

**Comment by stephengroat** [comment]
addressed the `http`/`https` issue

File: Formula/spotbugs.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
+class Spotbugs < Formula
+  desc "Tool for Java static analysis (FindBugs's successor)"
+  homepage "https://spotbugs.github.io/"
+  url "https://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz"
+  sha256 "c1eab4dc50cd9854c83fc4a90b4a9b9c63e5307bd0f0f7bb04720e17153b629c"
+  head "https://github.com/spotbugs/spotbugs.git"
+
+  bottle :unneeded
+
+  depends_on "gradle" => :build
```
Comment (DomT4) [inline Formula/spotbugs.rb]: Should this not be in a `head do` block looking at the below code?

File: Formula/spotbugs.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
+class Spotbugs < Formula
+  desc "Tool for Java static analysis (FindBugs's successor)"
+  homepage "https://spotbugs.github.io/"
+  url "https://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz"
+  sha256 "c1eab4dc50cd9854c83fc4a90b4a9b9c63e5307bd0f0f7bb04720e17153b629c"
+  head do
```
Comment (DomT4) [inline Formula/spotbugs.rb]: New line needed between the `sha256` and `head do`.

File: Formula/spotbugs.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
+class Spotbugs < Formula
+  desc "Tool for Java static analysis (FindBugs's successor)"
+  homepage "https://spotbugs.github.io/"
+  url "https://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.0/spotbugs-3.1.0.tgz"
+  sha256 "c1eab4dc50cd9854c83fc4a90b4a9b9c63e5307bd0f0f7bb04720e17153b629c"
+
+  head do
+    url "https://github.com/spotbugs/spotbugs.git"
```
Comment (ilovezfs) [inline Formula/spotbugs.rb]: need a blank line

Comment (stephengroat) [reply Formula/spotbugs.rb]: between `url` and `depends_on` inside of the `head` block? sorry, i modeled after `libp11.rb`

**Comment by stephengroat** [comment]
@BrewTestBot test this please

File: Formula/spotbugs.rb (hunk 13 above)
Comment (ilovezfs) [reply Formula/spotbugs.rb]: yes between url and depends_on.

Comment (stephengroat) [reply Formula/spotbugs.rb]: Sorry about that, done

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by stephengroat** [comment]
no pressure, just let me know if there's anything else i need to do for this

**Comment by ilovezfs** [comment]
Thanks for another great contribution to Homebrew @stephengroat!

**Pull request merged**